finat/fiat_elements.py

```diff
@@ -190,103 +190,50 @@ def dual_basis(self):
-                # get weights q_j for point x_j from complicated FIAT dual data
-                # structure
-                weight_dict = {cmp: weight for weight, cmp in tuples}
-                if len(self._element.value_shape()) == 0:
-                    q_j = np.array(weight_dict[tuple()])
-                else:
-                    q_j = np.zeros(self._element.value_shape())
-                    for key, item in weight_dict.items():
-                        q_j[key] = item
-
-                # Ensure all weights have the same shape
-                if last_shape is not None:
-                    assert q_j.shape == last_shape
-                last_shape = q_j.shape
-
-                # Create key into x dict
-                x_key = np.round(x_j.points, x_key_decimals)
-                assert np.allclose(x_j.points, x_key, atol=x_key_atol)
-                x_key = tuple(x_key.flatten())
-
-                # Get value and index k or add to dict. k are the columns of Q.
-                try:
-                    x_j, k = x[x_key]
-                except KeyError:
-                    k = len(x)
-                    x[x_key] = x_j, k
-
-                # q_j may be tensor valued
-                it = np.nditer(q_j, flags=['multi_index'])
-                for q_j_entry in it:
-                    Q[(i, k) + it.multi_index] = q_j_entry
-                    if len(set((i, k) + it.multi_index)) > 1:
-                        # Identity has i == k == it.multi_index[0] == ...
-                        # Since i increases from 0 in increments of 1 we know
-                        # that if this check is not triggered we definitely
-                        # have an identity tensor.
-                        self.Q_is_identity = False
-
-        #
-        # CONVERT Q TO gem.Literal (TODO: should be a sparse tensor)
-        #
-
-        # temporary until sparse literals are implemented in GEM which will
-        # automatically convert a dictionary of keys internally.
-        Q = gem.Literal(sparse.as_coo(Q).todense())
-        #
-        # CONVERT x TO gem.PointSet
-        #
-
-        # Convert PointSets to a single PointSet with the correct ordering
-        # for contraction with Q
-        random_pt, _ = next(iter(x.values()))
-        dim = random_pt.dimension
-        allpts = np.empty((len(x), dim), dtype=random_pt.points.dtype)
-        for _, (x_k, k) in x.items():
-            assert x_k.dimension == dim
-            allpts[k, :] = x_k.points
-        assert allpts.shape[1] == dim
+            pts = tuple(sorted(point_dict.keys()))
+            kstart, kend = seen[pts]
+            for p, k in zip(pts, range(kstart, kend)):
+                for weight, cmp in point_dict[p]:
+                    Q[(i, k, *cmp)] = weight
+
         x = PointSet(allpts)
+        if all(len(set(key)) == 1 and np.isclose(weight, 1) and len(key) == 2
+               for key, weight in Q.items()):
+            # Identity, is this the most general case?
```

Comment:
No. You are looking purely for an identity matrix whereas my existing check would find any identity-like tensor I that acted on a tensor T to produce T.

Reply:
The `all(len(set(key)) == 1` is fine but `np.isclose(weight, 1) and len(key) == 2` are too restrictive. 

I think we get an identity tensor `Q` of shape `(X, X, X)` (say) indexed as `Q[i, j, k]` whenever we have 
```python
np.isclose(Q[i, j, k], 1) and i==j==k
```

Reply:
I'm not sure yet what that looks like in terms of kronecker deltas

Reply:
I guess it's probably something like an outer product of a bunch of Deltas but I'm not sure (and google isn't being helpful). This is why i just flagged this case in the other code and didn't do the contraction when I identified identity

Reply:
So I want to understand exactly what you mean by "Identity" here. Which will probably help me.

I think that you have a pointset (with shape `(p,)`), and evaluate the function (on an expression which has shape `(q1, q2)` say) to get something with shape `(p, q1, q2)`. Now you have a Q tensor with shape `(n, p, q1, q2)` and you want that:

```
CT(sum(Q[i, j, k, l]*expr[j, k, l], (j, k, l)), (i,)) == expr
```
?

But this seems not right because the left hand side has shape `(n,)` and the right hand side has shape `(p, q1, q2)`

Reply:
I suppose this only works at least when `p == n` (which is the case if all the nodes are point evaluations?)

Reply:
it's the same optimisation that already existed in the TSFC dual eval code for point evaluations, yes

Reply:
What about the other part of the question?

Reply:
such that rather than doing any contraction you have that 
```
evaluation_expr, basis_indices = element.dual_evaluation(fn)
```
has `basis_indices = x.indices` and `evaluation_expr = fn(x)` (where `x` is from `element.dual_basis`)

Reply:
I think this means that `evaluation_expr` must have shape `(p,) == (x.indices[0].extent,)` (with my example above), and so there will never be `q1` and `q2`?

Reply:
> What about the other part of the question?

By `Q` being identity I mean that `Q` doesn't actually do anything numerically to `expr = fn(x)` when contracted along `x.indices + expr_shape_indices`: once the code has run it would yield the same thing as `expr`

Reply:
I think that if I contract over the `expr_shape_indices` then my output expression must be of lower rank than `expr` (unless `expr_shape_indices == ()`), what am I missing?

Reply:
> I think this means that `evaluation_expr` must have shape `(p,) == (x.indices[0].extent,)` (with my example above), and so there will never be `q1` and `q2`?

Yes I think you are correct given your shape argument.

Reply:
I think there might be an equivalent optimisation in there for tensor point sets where you don't have expression shape to contract over but I'm not quite sure what it is yet and I certainly hadn't implemented it up to this point

Reply:
I believe that with the implementation via deltas we have here, this "just works", at least, the code comes out looking right.

This is the kernel on a quad interpolate Q1 into Q2 if I replace the `Delta` with just a tabulated `Q` for the factors:
```
#include <stdint.h>

void expression_kernel(double *__restrict__ A, double const *__restrict__ w_0)
{
  double const t0[3 * 3] = { 1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0 };
  double const t1[3 * 2] = { 1.0, 0.0, 0.5, 0.5, 0.0, 1.0 };
  double t2[3 * 2];
  double t3[2];
  double t4[2];

  for (int32_t p1 = 0; p1 <= 2; ++p1)
  {
    for (int32_t i = 0; i <= 1; ++i)
      t2[2 * p1 + i] = 0.0;
    for (int32_t i_0 = 0; i_0 <= 2; ++i_0)
      for (int32_t i_1 = 0; i_1 <= 1; ++i_1)
        t2[2 * p1 + i_1] = t2[2 * p1 + i_1] + t0[3 * p1 + i_0] * t1[2 * i_0 + i_1];
  }
  for (int32_t p0 = 0; p0 <= 2; ++p0)
  {
    for (int32_t i_2 = 0; i_2 <= 1; ++i_2)
    {
      t3[i_2] = 0.0;
      for (int32_t i_3 = 0; i_3 <= 2; ++i_3)
        t3[i_2] = t3[i_2] + t0[3 * p0 + i_3] * t1[i_2 + 2 * i_3];
    }
    for (int32_t i_4 = 0; i_4 <= 1; ++i_4)
    {
      t4[i_4] = 0.0;
      for (int32_t i_5 = 0; i_5 <= 1; ++i_5)
        t4[i_4] = t4[i_4] + w_0[i_4 + 2 * i_5] * t3[i_5];
    }
    for (int32_t p1_0 = 0; p1_0 <= 2; ++p1_0)
      for (int32_t i_6 = 0; i_6 <= 1; ++i_6)
        A[3 * p0 + p1_0] = A[3 * p0 + p1_0] + t2[2 * p1_0 + i_6] * t4[i_6];
  }
}
```

And this is the kernel with the deltas left in:

```
void expression_kernel(double *__restrict__ A, double const *__restrict__ w_0)
{
  double const t0[3 * 2] = { 1.0, 0.0, 0.5, 0.5, 0.0, 1.0 };
  double t1[2];

  for (int32_t p0 = 0; p0 <= 2; ++p0)
  {
    for (int32_t i = 0; i <= 1; ++i)
    {
      t1[i] = 0.0;
      for (int32_t i_0 = 0; i_0 <= 1; ++i_0)
        t1[i] = t1[i] + t0[2 * p0 + i_0] * w_0[i + 2 * i_0];
    }
    for (int32_t p1 = 0; p1 <= 2; ++p1)
      for (int32_t i_1 = 0; i_1 <= 1; ++i_1)
        A[3 * p0 + p1] = A[3 * p0 + p1] + t0[2 * p1 + i_1] * t1[i_1];
  }
}
```

(The former has a flop count of 132, the latter 60)

Reply:
Note how all the multiplications by the identity matrix have disappeared.

Reply:
Excellent. I have also notices a flop count decrease on the CG1->RT2 script I tested in the firedrake PR (looks like better loop nesting) and general performance improvements all round

Reply:
How do we test if sum factorisation is actually working?

Reply:
If we merge firedrakeproject/tsfc#256 (or cherry pick https://github.com/firedrakeproject/tsfc/pull/256/commits/496b32460d7c796e2decb41f4a703abf730b196f from that branch, since the complex version of the tests is still broken so the merge is blocked), then when the kernel you get back from the `c_e_d_e` call has a `flop_count` property. On tensor product cells for Q and DQ spaces, these flop counts should asymptotically go like `p^{d+1}` (when interpolating from (say) Q_{p-1} to Q_{p}). That's for simple interpolation of a coefficient, you can do the same thing for interpolation of more complicated expressions.

Reply:
And what's a nice everyday example of a tensor product element someone might be using?

Reply:
`FiniteElement("Q", hexahedron, p)`